src/main/scala/units/ELUpdater.scala

```diff
@@ -1474,17 +1511,43 @@ class ELUpdater(
         }
       })
 
-    (depositedTransactions, addedAssets, updateAssetRegistryTransaction, nativeTransfersViaDeposits, assetTransfers)
+    val depositedTransactions = updateAssetRegistryTransaction.toVector ++ transferTransactions
+
+    (depositedTransactions, addedAssets, updateAssetRegistryTransaction, transferTransactions, nativeTransfersViaDeposits, assetTransfers)
+  }
+
+  private def getFailedTransfers(
+      c2eTransferLogs: List[GetLogsResponseEntry],
+      transfersWithTransactions: Seq[(ContractTransfer, DepositedTransaction)]
+  ): Seq[ContractTransfer.Asset | ContractTransfer.NativeViaDeposit] = {
+    val successfulTransferHashes = c2eTransferLogs.map(_.transactionHash).toSet
+    transfersWithTransactions.collect {
+      case (transfer: (ContractTransfer.Asset | ContractTransfer.NativeViaDeposit), dt) if !successfulTransferHashes.contains(dt.hash) =>
+        transfer
+    }
   }
 
   private def validateC2ETransfers(
       actualWithdrawals: Seq[Withdrawal],
       actualTransferLogs: List[GetLogsResponseEntry],
       expectedTransfers: Seq[ContractTransfer],
+      transferTransactions: Seq[DepositedTransaction],
       prevWithdrawalIndex: Long,
-      strictC2ETransfersActivated: Boolean
+      strictC2ETransfersActivated: Boolean,
+      expectedFailedC2ETransfersRootHash: Digest,
+      blockHash: BlockHash
   ): Either[String, Long] = {
-    val totalTransfers = expectedTransfers.size
+    val totalTransfers                = expectedTransfers.size
+    val failedTransfers               = getFailedTransfers(actualTransferLogs, expectedTransfers.zip(transferTransactions))
+    val actualFailedTransfersRootHash = BridgeMerkleTree.getFailedTransfersRootHash(failedTransfers.map(_.index))
+
+    val failedC2ETransfersRootHashCheck =
+      Either.raiseUnless(java.util.Arrays.equals(actualFailedTransfersRootHash, expectedFailedC2ETransfersRootHash)) {
+        s"Failed CL to EL transfers root hash mismatch in block ${blockHash}: "
+          + s"EL=${toHexNoPrefix(actualFailedTransfersRootHash)}, " +
+          s"CL=${toHexNoPrefix(expectedFailedC2ETransfersRootHash)}"
```

Comment:
```suggestion
      Either.raiseUnless(java.util.Arrays.equals(actualFailedTransfersRootHash, expectedFailedC2ETransfersRootHash)) {
        s"Failed CL to EL transfers root hash mismatch in block $blockHash: " +
          s"EL=${toHexNoPrefix(actualFailedTransfersRootHash)}, " +
          s"CL=${toHexNoPrefix(expectedFailedC2ETransfersRootHash)}"
      }
```
IntelliJ IDEA tells, the first line is not used. I'm not sure this is true, but let's stick to a consistent style :)